Framework/CurveFitting/src/Functions/Convolution.cpp

```diff
@@ -270,7 +297,132 @@ void Convolution::function(const FunctionDomain &domain,
       std::transform(out, out + nData, tmp.data(), out, std::plus<double>());
     }
   }
-}
+
+} // end of Convolution::functionFFTMode
+
+/**
+ * Calculates convolution of the two member functions when the
+ * domain is not symmetric with respect to inversion E --> -E.
+ * @param domain :: space on which the function acts
+ * @param values :: buffer to store the values returned by the function after
+ * acting on the domain.
+ */
+void Convolution::functionDirectMode(const FunctionDomain &domain,
+                                     FunctionValues &values) const {
+  const FunctionDomain1D *d1d = dynamic_cast<const FunctionDomain1D *>(&domain);
+  const size_t nData = domain.size();
+  const double *xValues = d1d->getPointerAt(0);
+  double dx =
+      (xValues[nData - 1] - xValues[0]) / static_cast<double>((nData - 1));
+  auto ixP = static_cast<size_t>(xValues[nData - 1] / dx); // positive
+                                                           // x-values
+  auto ixN = nData - ixP - 1; // negative x-values (ixP+ixN=nData-1)
+
+  refreshResolution();
+
+  // double the domain where to evaluate the convolution. Guarantees complete
+  // overlap betwen convolution and signal in the original range.
+  const size_t mData = nData + ixN + ixP; // equal to 2*nData-1
+  std::vector<double> xValuesExtdVec(mData);
+  double *xValuesExtd = xValuesExtdVec.data();
+  Mantid::API::FunctionDomain1DView domainExtd(xValuesExtd, mData);
+  double Dx = dx * static_cast<double>(ixN + ixP);
+  for (size_t i = 0; i < mData; i++) {
+    xValuesExtd[i] = -Dx + static_cast<double>(i) * dx;
+  }
+
+  // Fill m_resolution with the resolution function data
+  IFunction1D_sptr resolution =
+      boost::dynamic_pointer_cast<IFunction1D>(getFunction(0));
+  if (!resolution) {
+    throw std::runtime_error("Convolution can work only with IFunction1D");
+  }
+  if (m_resolution.empty()) {
+    m_resolution.resize(nData);
+  }
+  resolution->function1D(m_resolution.data(), xValues, nData);
+
+  // Reverse the axis of the resolution data
+  std::reverse(m_resolution.begin(), m_resolution.end());
+
+  // check for delta functions
+  std::vector<boost::shared_ptr<DeltaFunction>> dltFuns;
+  double dltF = 0;
```

Comment:
Code dealing with the delta functions in this method duplicates code from functionFFTMode. Can this code be moved to a separate method(s) or standalone function(s)?